### ISSUE TITLE
`borrowed_box` incorrectly assumes mutable boxed slice allocation won't change

### ISSUE THREAD

**Issue by okready** [rust-lang/rust-clippy#2907]
Defining a function that has a `&mut Box<[T]>` parameter triggers the `borrowed_box` lint, recommending it be replaced with a `&mut [T]`. It's possible to have a function that replaces the boxed slice with a different allocation (e.g. a slice with a different size), which cannot be done if the parameter is just a slice reference.

Example:

```rust
pub fn maybe_sets_box(boxed_slice: &mut Box<[i32]>) {
    if boxed_slice.is_empty() {
        let mut data = Vec::with_capacity(1);
        data.push(12);
        *boxed_slice = data.into_boxed_slice();
    }
}
```

**Comment by EFanZh** [comment]
I think Clippy should allow the usage of `&mut Box<T>` but disallow the usage of `&Box<T>` since we might want to change the box pointer instead of the object it points to.

**Comment by jmaargh** [comment]
A (perhaps a little contrived) example: https://godbolt.org/z/LFRvd7

**Comment by ebroto** [comment]
I think this issue can be closed given that #5491 was merged
